src/main/java/org/prebid/server/vertx/http/CircuitBreakerSecuredHttpClient.java

```diff
@@ -0,0 +1,113 @@
+import org.prebid.server.metric.Metrics;
+
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Function;
+
+/**
+ * Wrapper over {@link HttpClient} with circuit breaker functionality.
+ */
+public class CircuitBreakerSecuredHttpClient implements HttpClient {
+
+    private static final Logger logger = LoggerFactory.getLogger(CircuitBreakerSecuredHttpClient.class);
+
+    private final HttpClient httpClient;
+    private final Metrics metrics;
+
+    private final Map<String, CircuitBreaker> circuitBreakerByName = new ConcurrentHashMap<>();
+    private final Function<String, CircuitBreaker> circuitBreakerCreator;
+
+    public CircuitBreakerSecuredHttpClient(Vertx vertx, HttpClient httpClient, Metrics metrics,
+                                           int maxFailures, long timeoutMs, long resetTimeoutMs) {
+        Objects.requireNonNull(vertx);
+        this.httpClient = Objects.requireNonNull(httpClient);
+        this.metrics = Objects.requireNonNull(metrics);
+
+        circuitBreakerCreator = name -> CircuitBreaker.create("http-client-circuit-breaker-" + name,
+                vertx,
+                new CircuitBreakerOptions()
+                        .setMaxFailures(maxFailures)
+                        .setTimeout(timeoutMs)
+                        .setResetTimeout(resetTimeoutMs))
+                .openHandler(ignored -> circuitOpened(name))
+                .halfOpenHandler(ignored -> circuitHalfOpened(name))
+                .closeHandler(ignored -> circuitClosed(name));
+    }
+
+    private void circuitOpened(String name) {
+        logger.warn(String.format("Http client request to %s is failed, circuit opened.", name));
+        metrics.updateHttpClientCircuitBreakerMetric(true);
+    }
+
+    private void circuitHalfOpened(String name) {
+        logger.warn(String.format("Http client request to %s will try again, circuit half-opened.", name));
+    }
+
+    private void circuitClosed(String name) {
+        logger.warn(String.format("Http client request to %s becomes succeeded, circuit closed.", name));
+        metrics.updateHttpClientCircuitBreakerMetric(false);
+    }
+
+    @Override
+    public void request(HttpMethod method, String url, MultiMap headers, String body, long timeoutMs,
+                        Handler<HttpClientResponse> responseHandler, Handler<Throwable> exceptionHandler) {
+
+        final String name = nameFrom(url);
+        final CircuitBreaker breaker = circuitBreakerByName.computeIfAbsent(name, circuitBreakerCreator);
+        circuitBreakerByName.putIfAbsent(name, breaker);
+
+        breaker.<HttpClientResponse>execute(future -> httpClient.request(method, url, headers, body, timeoutMs,
+                response -> handleResponse(response, future),
+                exception -> handleException(exception, future)))
+                .setHandler(ar -> {
```

Comment:
After discussion with Ruslan it was decided to try changing method signature to return `Future`.